### DISCUSSION TITLE
Handling non-linear non-differentiable state

### DISCUSSION THREAD

**Discussion by gioperin** [do-mpc/do-mpc#374] [Q&A] [answered]
Hello @4flixt and all,

I have an issue with a system that I am currently trying to simulate. It is related to the ODE of a discrete model simulating the level of a battery in a computing system. Let x be my state, my equation is `x_next = min{Ax + Bu + W, x_max}`. So, I am trying to encode this with casadi if-else condition:

`x_next = Ax + Bu + W`
`x_next = cs.if_else(x_next >= x_max, x_max, x_next)`.

However, this is not working: when I look in the future in `mpc.opt_x_num` my variable, it is not upper bounded by x_max.
Note that just using an upper bound for the state x is not a feasible solution, because it would affect the behavior of the control input u (which is CPU operations that are done by the computing system. I do not want my system to do useless CPU operations just because otherwise the battery overflows!) 

Any suggestion is more than welcome!

Best,

Giovanni

**Comment by 4flixt** [comment, marked as answer]
Hi @gioperin,

apologies for the late reply. We generally don't advise to use ``if_else`` or other non-differentiable expressions if you want to use the respective model in an MPC formulation. 
That being said, your formulation should work - at least for the simulator.
Alternatively, you can try:
```python
import casadi as cs

x_next_clipped = cs.fmin(x_next, x_max)
```
with [fmin](https://web.casadi.org/docs/#list-of-operations) computing the element-wise minimum according to CasADis documentation.

If you wish to  use your model in an MPC formulation, I strongly recommend you ditch this clipping formulation and instead use a constraint.
If there is a feasible solution that avoids ``x>x_max``, the controller will apply it to the system.
